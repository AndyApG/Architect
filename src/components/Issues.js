import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { map, size } from 'lodash';
import scrollparent from 'scrollparent';
import anime from 'animejs';
import { getCSSVariableAsObject, getCSSVariableAsNumber } from '../utils/CSSVariables';
import { flattenIssues, getFieldId } from '../utils/issues';

const scrollTo = (destination) => {
  if (!destination) { return; }
  const scroller = scrollparent(destination);
  const scrollStart = scroller.scrollTop;
  const destinationOffset = parseInt(destination.getBoundingClientRect().top, 10);
  const scrollEnd = scrollStart + destinationOffset;

  anime({
    targets: scroller,
    scrollTop: scrollEnd,
    easing: getCSSVariableAsObject('--animation-easing-js'),
    duration: getCSSVariableAsNumber('--animation-duration-fast-ms'),
  });
};

class Issues extends Component {
  static propTypes = {
    show: PropTypes.bool,
    issues: PropTypes.array,
  };

  static defaultProps = {
    show: true,
    issues: [],
  };

  constructor(props) {
    super(props);

    this.state = {
      open: true,
    };
  }

  componentWillReceiveProps(newProps) {
    const noIssues = size(newProps.issues) > 0;
    const show = newProps.show;

    // when panel hidden by parent reset collapsed state
    if (noIssues || !show) { this.setState({ open: true }); }
  }

  isVisible = () =>
    this.props.show && size(this.props.issues) > 0;

  isOpen = () =>
    this.state.open;

  handleClickTitleBar = () =>
    this.setState({ open: !this.state.open });

  handleClickIssue = (e) => {
    e.preventDefault();

    const link = e.target.getAttribute('href');
    const destination = document.querySelector(link);
    scrollTo(destination);
  }

  render() {
    const flatIssues = flattenIssues(this.props.issues);

    const issues = map(
      flatIssues,
      ({ field, issue }) => (
        <li key={getFieldId(field)} className="issues__issue">
          <a
            href={`#${getFieldId(field)}`}
            onClick={this.handleClickIssue}
          >
            {field} is {issue}
          </a>
        </li>
      ));

    const issuesClasses = cx(
      'issues',
      {
        'issues--show': this.isVisible(),
        'issues--open': this.isOpen(),
      },
    );

    return (
      <div className={issuesClasses}>
        <div className="issues__panel">
          <div className="issues__title-bar" onClick={this.handleClickTitleBar}>
            Issues ({size(this.props.issues)}) {this.isOpen()} { this.isVisible() }
          </div>
          <ol className="issues__issues">
            {issues}
          </ol>
        </div>
      </div>
    );
  }
}

export { Issues };

export default Issues;
