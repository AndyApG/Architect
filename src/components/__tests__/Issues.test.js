/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { Issues } from '../Issues';

const mockIssues = {
  foo: 'bar',
  baz: [
    {
      buzz: 'foo',
      beep: 'boop',
    },
  ],
};

const mockProps = {
  show: true,
  issues: {},
};

describe('<Issues />', () => {
  it('will render', () => {
    const component = shallow(<Issues {...mockProps} />);

    expect(component).toMatchSnapshot();
  });

  it('renders issues from object', () => {
    const component = shallow(<Issues show issues={mockIssues} />);
    expect(component.find('li.issues__issue').length).toBe(3);
  });
});
