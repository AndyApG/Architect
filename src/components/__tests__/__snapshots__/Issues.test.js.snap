// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Issues /> will render 1`] = `
<div
  className="issues issues--open"
>
  <div
    className="issues__panel"
  >
    <div
      className="issues__title-bar"
      onClick={[Function]}
    >
      <div
        className="issues__title-bar-icon"
      >
        <Icon
          className=""
          color="white"
          name="info"
          style={Object {}}
        />
      </div>
      <div
        className="issues__title-bar-text"
      >
        Issues (
        0
        )
      </div>
      <div
        className="issues__title-bar-toggle"
      >
        <Icon
          className="issues__toggle--open"
          color="white"
          name="chevron-down"
          style={Object {}}
        />
        <Icon
          className="issues__toggle--close"
          color="white"
          name="chevron-up"
          style={Object {}}
        />
      </div>
    </div>
    <ol
      className="issues__issues"
    />
  </div>
</div>
`;
